k2/csrc/top_sort.cu

```diff
@@ -91,15 +96,58 @@ FsaVec RenumberFsaVec(FsaVec &fsas, const Array1<int32_t> &order,
             fsas_src_idx01 = fsas_idx0x + fsas_src_idx1,
             fsas_dest_idx01 = fsas_idx0x + fsas_dest_idx1;
     K2_CHECK_EQ(old2new_data[fsas_src_idx01], ans_idx01);
-    int32_t ans_dest_idx01 = old2new_data[fsas_dest_idx01],
-            ans_dest_idx1 = ans_dest_idx01 - ans_idx0x;
-    arc.src_state = ans_idx1;
-    arc.dest_state = ans_dest_idx1;
-    ans_arcs_data[ans_idx012] = arc;
-    if (arc_map_data) arc_map_data[ans_idx012] = fsas_idx012;
+    int32_t ans_dest_idx01 = old2new_data[fsas_dest_idx01];
+    if (ans_dest_idx01 != -1) {
+      int32_t ans_dest_idx1 = ans_dest_idx01 - ans_idx0x;
+      arc.src_state = ans_idx1;
+      arc.dest_state = ans_dest_idx1;
+      ans_arcs_data[ans_idx012] = arc;
+      keep_data[ans_idx012] = 1;
+    } else {
+      keep_data[ans_idx012] = 0;
+    }
+    if (arc_map_data != nullptr) arc_map_data[ans_idx012] = fsas_idx012;
   };
-  Eval(c, ans_shape.NumElements(), lambda_set_arcs);
-  return FsaVec(ans_shape, ans_arcs);
+  Eval(c, ans_num_arcs, lambda_set_arcs);
+
+  if (order.Dim() != fsas.TotSize(1)) {
+    Array1<int32_t> valid_arc_new2old = arcs_renumbering.New2Old();
+    int32_t valid_num_arcs = valid_arc_new2old.Dim();
+    Array1<int32_t> valid_row_ids2(c, valid_num_arcs);
+    Array1<Arc> valid_arcs(c, valid_num_arcs);
+    int32_t *valid_row_ids2_data = valid_row_ids2.Data();
+    Arc *valid_arcs_data = valid_arcs.Data();
+    const int32_t *valid_arc_new2old_data = valid_arc_new2old.Data();
+
+    Array1<int32_t> valid_arc_map;
+    int32_t *valid_arc_map_data;
+    if (arc_map != nullptr) {
+      valid_arc_map = Array1<int32_t>(c, valid_num_arcs);
+      valid_arc_map_data = valid_arc_map.Data();
+    } else {
+      valid_arc_map_data = nullptr;
+    }
+
+    auto lambda_set_valid_arcs_and_row_ids2 =
+        [=] __host__ __device__(int32_t valid_arc_idx012) -> void {
+      int32_t ans_arc_idx012 = valid_arc_new2old_data[valid_arc_idx012];
+      valid_row_ids2_data[valid_arc_idx012] = ans_row_ids2_data[ans_arc_idx012];
+      valid_arcs_data[valid_arc_idx012] = ans_arcs_data[ans_arc_idx012];
+      if (arc_map_data != nullptr)
+        valid_arc_map_data[valid_arc_idx012] = arc_map_data[ans_arc_idx012];
+    };
+    Eval(c, valid_num_arcs, lambda_set_valid_arcs_and_row_ids2);
+
+    Array1<int32_t> valid_row_splits2(c, new_num_states + 1);
+    RowIdsToRowSplits(valid_row_ids2, &valid_row_splits2);
```

Comment:
@danpovey , I want to call `SumPerSublist` here on `keep_data` to get `valid_row_splits2` (in another stream), but the type of `keep_data` is `Array1<char>` and  the type of `row_splits` is `Array1<int32_t>`. So from a perspective of long term usage, should we change the type of `keep` in `Renumbering` to `Array1<int32_t>`? or should we support `ApplyOpPerSublist` with a template `dest` type, e.g.
```
template <typename T, typename S, typename Op>
void ApplyOpPerSublist(Ragged<T> &src, T default_value, Array1<S> *dst);
```
I'm not sure about which is better for general usage in a long term.

Reply:
I'm open to changing the type of `keep`, and possible even merging it with the old2new_ vector (we can do the summation in-place).  But let's not do that now.
I don't think it should be necessary to do the sum per sub-list.  Can't you just take differences between the elements of the exclusive-sum?

Reply:
Of course we can take the differences, but that also needs we do exclusive-sum then take differences (two kernels).

So I'll keep the current code as-is, let's update the type of `keep` in the future.

Reply:
Something seems to me not right about what you are saying.  A row-splits vector would naturally come from an exclusive-sum and then taking certain elements...  taking differences doesn't seem right.